src/spi.rs

```diff
@@ -709,3 +804,132 @@ impl_spi!(SPI3: spi_host_device_t_SPI3_HOST);
 impl_spi_any_pins!(SPI2);
 #[cfg(not(esp32c3))]
 impl_spi_any_pins!(SPI3);
+
+use crate::gpio::{Level, Output};
+use crate::task::CriticalSection;
+use crate::delay::Ets;
+
+pub struct GuardedSpiDevice<'d, DRIVER: Borrow<SpiMasterDriver<'d>>> {
+    device: EspSpiDevice<'d, DRIVER>,
+    mutex: CriticalSection,
+}
+
+impl<'d, DRIVER> GuardedSpiDevice<'d, DRIVER>
+where
+    DRIVER: Borrow<SpiMasterDriver<'d>> + 'd,
+{
+    pub fn new(device: EspSpiDevice<'d, DRIVER>) -> Self {
+        Self {
+            device,
+            mutex: CriticalSection::new(),
+        }
+    }
+}
+pub struct EspSCSSpiDevice<'d, DEVICE, DRIVER: Borrow<SpiMasterDriver<'d>>> {
+    shared_device: DEVICE,
+    cs_pin: PinDriver<'d, AnyOutputPin, Output>,
+    pre_delay: Option<u32>,
+    post_delay: Option<u32>,
+    _p: PhantomData<&'d DRIVER>,
```

Comment:
Remove `_p`, it does not serve any purpose in this context. And it will allow you to:
a) Get rid of the `DRIVER` generic on the struct definition level
b) rename `DEVICE` to `DRIVER`

Reply:
i had a hard time here. i implemented the lock Fn to make use of the interior mutability, bit i cant get ride of these so easy. you can have a look at the updated version. advice appreciated

Reply:
Surely there's some sort of `CriticalSection<T>` utility somewhere.

Yeah I don't see how the generic params could be removed either. Their kinda needed to borrow the shared driver.

Reply:
Yeah. Seems not possible. I'll try post-merge but very unlikely to workl.